2018 Fantasy Football API Update

Robert-litts opened the pull request:
Took another look at the code and made a second attempt to fix #514. The issue is caused by lines 48-51 in espn_requests.py, where the endpoint used for 2018 league in my case should be the older endpoint. When that endpoint is used, the data returned has a 0 index and needs to be handled appropriately. Updates made in base_league.py to handle the 0 index JSON data. Additionally, created a new function in epsn_requests.py to validate if endpoint causes a 401 and if so, try the other endpoint, raise an error if both fail.

Robert-litts commented:
This also addresses my issue #278.

cwendt94 commented on espn_api/requests/espn_requests.py:
```diff
@@ -73,6 +75,10 @@ def get_league(self):
         }
         data = self.league_get(params=params)
+        # Check if the data is a list (which happens when year is 2018 or earlier)
+        if isinstance(data, list):
```
this check can be moved to league_get call so that it only has to be done once. league_get is already doing this for any year before 2018. That logic can be updated with this check instead

Robert-litts replied:
This makes sense, I missed that the first go around. Updated the logic there & removed the check for league year.

Robert-litts commented:
Made the changes & tested it against my previously failing 2018 league. Updates worked for all league years (2017 and prior, 2018, and 2019 and newer). Thanks for the suggestion on where to add this check into the code, and hope the changes make sense.

Robert-litts commented:
I saw the original tests failed because of the urllib3 version issue. I reran the tests for this PR after you updated/pinned that version earlier today and got the same errors on the test cases. From my review so far, it looks like because of where I implemented the "check_league_endpoint" function within the class __init__ it is being missed by the test cases. 

Taking another look tonight and I'm reconsidering the implementation and this might be more appropriate to place this logic within your existing checkRequestStatus function, which is already being run in both get/league_get.

Robert-litts commented:
OK, hopefully fourth time is a charm!

- Moved the alternate endpoint code/logic into checkRequestStatus function & slightly modified the league_get to handle the potential for an alternate endpoint. 

- Ran all nosetests in my dev container & all passed
- Did some testing with my original 2018 league that was causing issues and all seems to be working great.

cwendt94 commented on espn_api/requests/espn_requests.py:
```diff
@@ -17,19 +17,49 @@ class ESPNUnknownError(Exception):
 
 
-def checkRequestStatus(status: int, cookies=None, league_id=None) -> None:
+def checkRequestStatus(status: int, cookies=None, league_id=None, league_endpoint=None, year=None, extend="", params=None, headers=None) -> tuple:
```
To make this function a little simpler, I think moving this inside the request class so then you won't have to pass all these fields to it. You then can access all of them through `this.`

Robert-litts replied:
Yes I was actually considering that last night, but wasn't sure if you had a reason it was defined outside the class. Moving it inside would definitely simplify access to those params for sure.

cwendt94 replied:
Yeah no issues moving it in!

cwendt94 reviewed (commented):
Thanks for working on this! I think its really close!

Robert-litts commented:
Excellent -- I was experimenting with both your League(368876, 2018) and my League(747376, 2018) this morning and it seems like both of those are accessible at the 'leagueHistory' endpoint. So in your [test_private_league](https://github.com/cwendt94/espn-api/blob/ae12d005fa579907388bf9434adc9d2b51e2b2bf/tests/football/integration/test_league.py#L17), it will now not raise an exception with this code, causing the test to fail. At least I think that is what is happening, are you able to validate that assumption?

cwendt94 commented:
> Excellent -- I was experimenting with both your League(368876, 2018) and my League(747376, 2018) this morning and it seems like both of those are accessible at the 'leagueHistory' endpoint. So in your [test_private_league](https://github.com/cwendt94/espn-api/blob/ae12d005fa579907388bf9434adc9d2b51e2b2bf/tests/football/integration/test_league.py#L17), it will now not raise an exception with this code, causing the test to fail. At least I think that is what is happening, are you able to validate that assumption?

Wow is it really? Thats an old private league of mine and wouldn't think it would be accessible. I wonder if we update to an older year it would throw?

Robert-litts commented:
> > Excellent -- I was experimenting with both your League(368876, 2018) and my League(747376, 2018) this morning and it seems like both of those are accessible at the 'leagueHistory' endpoint. So in your [test_private_league](https://github.com/cwendt94/espn-api/blob/ae12d005fa579907388bf9434adc9d2b51e2b2bf/tests/football/integration/test_league.py#L17), it will now not raise an exception with this code, causing the test to fail. At least I think that is what is happening, are you able to validate that assumption?
> 
> Wow is it really? Thats an old private league of mine and wouldn't think it would be accessible. I wonder if we update to an older year it would throw?

Yeah my league is private too and I can access yours and mine without cookies, for all of the older years (mine goes back to 2013). However, I know this original issue was sporadic on some leagues, so I'm hesitant to say that a simple change to make the endpoint for 2018 always use the older URL would work for everyone.

cwendt94 commented:
> > > Excellent -- I was experimenting with both your League(368876, 2018) and my League(747376, 2018) this morning and it seems like both of those are accessible at the 'leagueHistory' endpoint. So in your [test_private_league](https://github.com/cwendt94/espn-api/blob/ae12d005fa579907388bf9434adc9d2b51e2b2bf/tests/football/integration/test_league.py#L17), it will now not raise an exception with this code, causing the test to fail. At least I think that is what is happening, are you able to validate that assumption?
> > 
> > 
> > Wow is it really? Thats an old private league of mine and wouldn't think it would be accessible. I wonder if we update to an older year it would throw?
> 
> Yeah my league is private too and I can access yours and mine without cookies, for all of the older years (mine goes back to 2013). However, I know this original issue was sporadic on some leagues, so I'm hesitant to say that a simple change to make the endpoint for 2018 always use the older URL would work for everyone.

Nice, well thats a little added feature. Definitely a weird bug on ESPN's side but glad you found a fix for it!

Robert-litts commented:
Ok I moved the checkRequestStatus function into the class. Everything seems to be working on my end & ran all the tests, and it is still failing just the one I referenced [above](https://github.com/cwendt94/espn-api/pull/615#issuecomment-2568627183) for your 2018 private league. 

How do you want me to handle that one test case failure? I'm not sure how to test that private league feature now since it doesn't appear to be a restriction w/ the data I have.

cwendt94 commented:
> 
> How do you want me to handle that one test case failure? I'm not sure how to test that private league feature now since it doesn't appear to be a restriction w/ the data I have.

I think its fine to remove that test

Robert-litts commented:
> > How do you want me to handle that one test case failure? I'm not sure how to test that private league feature now since it doesn't appear to be a restriction w/ the data I have.
> 
> I think we can use that test to get coverage on the switching endpoint logic. It can be updated to not throw error on private league.

I modified the test_private_league to validate the URLit switches to (i.e. it falls back to the `/leagueHistory/` endpoint), since per the code a year=2018 request should use the ` /seasons/` endpoint. I added random, invalid credentials to ensure it fails & falls back to the leagueHistory endpoint (but it should fail even without those). 

Interesting side-note, it appears the  `/leagueHistory/` no longer requires credentials -- I did some testing on some randomly generated league numbers and it looks like it succeeds every time.

cwendt94 commented:
> Interesting side-note, it appears the `/leagueHistory/` no longer requires credentials -- I did some testing on some randomly generated league numbers and it looks like it succeeds every time.

This looks good! Interesting, thats kind of nice!
